algebras/algebra/src/main/scala/endpoints4s/algebra/Middlewares.scala

```diff
@@ -0,0 +1,85 @@
+package endpoints4s.algebra
+
+import endpoints4s._
+
+/** Algebra for operations that a middleware can perform on requests.
+  *
+  * @see [[Middlewares]]
+  */
+trait RequestMiddlewares extends Requests {
+
+  def addRequestHeaders[A, H](
+      request: Request[A],
+      headers: RequestHeaders[H]
+  )(implicit tupler: Tupler[A, H]): Request[tupler.Out]
+
+  def addRequestQueryString[A, Q, Out](
+      request: Request[A],
+      qs: QueryString[Q]
+  )(implicit tupler: Tupler[A, Q]): Request[tupler.Out]
+
+  implicit class RequestMiddlewareOps[A](request: Request[A]) {
+
+    def withHeaders[H](headers: RequestHeaders[H])(implicit
+        tupler: Tupler[A, H]
+    ): Request[tupler.Out] =
+      addRequestHeaders(request, headers)
+
+    def withQueryString[Q](qs: QueryString[Q])(implicit
+        tupler: Tupler[A, Q]
+    ): Request[tupler.Out] =
+      addRequestQueryString(request, qs)
+  }
+}
+
+/** Algebra for operations that a middleware can perform on responses.
+  *
+  * @see [[Middlewares]]
+  */
+trait ResponseMiddlewares extends Responses { this: Errors =>
+
+  def addResponseHeaders[A, H](
+      response: Response[A],
+      headers: ResponseHeaders[H]
+  )(implicit tupler: Tupler[A, H]): Response[tupler.Out]
+
+  implicit class ResponseMiddlewareOps[A](response: Response[A]) {
+
+    def withHeaders[H](headers: ResponseHeaders[H])(implicit
+        tupler: Tupler[A, H]
+    ): Response[tupler.Out] =
+      addResponseHeaders(response, headers)
+  }
+}
+
+/** Algebra to modify an endpoint.
+  *
+  * This can be used for defining an Authentication or tracing middleware for example.
+  */
+trait Middlewares
+    extends EndpointsWithCustomErrors
+    with RequestMiddlewares
+    with ResponseMiddlewares {
+
+  def mapEndpointRequest[A, B, C](
+      endpoint: Endpoint[A, B],
+      f: Request[A] => Request[C]
+  ): Endpoint[C, B]
+  def mapEndpointResponse[A, B, C](
+      endpoint: Endpoint[A, B],
+      f: Response[B] => Response[C]
+  ): Endpoint[A, C]
+  def mapEndpointDocs[A, B](
+      endpoint: Endpoint[A, B],
+      f: EndpointDocs => EndpointDocs
+  ): Endpoint[A, B]
+
+  implicit class EndpointMiddlewareOps[A, B](endpoint: Endpoint[A, B]) {
+    def mapRequest[C](f: Request[A] => Request[C]): Endpoint[C, B] =
+      mapEndpointRequest(endpoint, f)
+    def mapResponse[C](f: Response[B] => Response[C]): Endpoint[A, C] =
+      mapEndpointResponse(endpoint, f)
+    def mapDocs(f: EndpointDocs => EndpointDocs): Endpoint[A, B] =
+      mapEndpointDocs(endpoint, f)
+  }
```

Comment:
I'm not sure I see how adding accessors helps with #756. 

Currenlty the `Endpoint` concrete types for client and server interpreters do not keep a reference to the `EndpointDocs` because it isn't used in anyway. So for these interpreters, `mapDocs` is a no-op. Only the openapi interpreter actually keeps a reference to `EndpointDocs`. So if we add accessors in the algebra, all interpreters would need to keep the reference to `EndpointDocs`.

Reply:
Sorry, I should've been more clear. I didn't mean an accessor for `EndpointDocs`, but for other parts of the endpoint, such as the request url, the response statuscode etc. I can try it out in a separate PR or even as part of #756 if that's preferred, so it doesn't hold back this PR.